File: C/Polymul.cpp
#include <iostream>
#include <vector>
#include <complex>
#include <cmath>

#define PI 3.14159265358979323846L

typedef std::complex <long double> V;
typedef std::vector<V> S;

long long pad (int l){
    l -= 1;
    int counter = 0;
    while (l!=0){
        l >>= 1;
        counter += 1;
    }
    return 1LL<<counter;
}

S fft(S& a, int m, V w){
    if (m == 1){
        return S(a);
    }
    S ae, ao;
    for (int i = 0; i < m/2; i++){
        ae.push_back(a[2*i]);
        ao.push_back(a[2*i+1]);
    }
    S fe, fo, F(m,0);
    fe = fft (ae, m/2, w*w);
    fo = fft (ao, m/2, w*w);
    V x = 1.0;
    for (int j=0; j< m/2; j++){
        F[j] = fe[j] + x*fo[j];
        F[j+m/2] = fe[j] - x*fo[j];
        x *= w;
    }
    return F;
}

void code(){
    int a, b,c=0;
    S Q, R, P, newd1, newd2, outputlist;
    std::cin >> a;
    while (c++ <= a){
        long double x;
        std::cin >> x;
        Q.push_back(x);
    }
    c = 0;
    std::cin >> b;
    while (c++ <= b){
        long double x;
        std::cin >> x;
        R.push_back(x);
    }
    int piss = a + b + 1;
    long long variable = pad(piss);
    while (Q.size() < variable){
        Q.push_back(0);
    }
    while (R.size() < variable){
        R.push_back(0);
    }
    V w(cos(2*PI/variable), sin(2*PI/variable));
    newd1 = fft(Q, variable, w);
    newd2 = fft(R, variable, w);
    for (int i=0; i<variable; i++){
        outputlist.push_back(newd1[i]*newd2[i]);
    }
    P = fft (outputlist, variable, V(1.0)/w);
    std::cout << piss-1 << "\n";
    for (int i=0; i< piss; i++){
        long double b = 0.5 + P[i].real()/variable;
        if (b<0){
            b -= 1;
        }
        std::cout << (int)b << " ";
    }
}

int main (){
    int N;
    std::cin >> N;
    while (N--){
        code();
        std::cout << "\n";
    }

    return 0;
}

File: Python/Polymul.py
import math

def pad (l):
    l -= 1
    counter = 0
    while l != 0:
        l = l >> 1
        counter += 1
    return 2**counter

def fft (a,m,w):
    if m == 1:
        return a
    ae = [a[2*i] for i in range (m/2)]     
    ao = [a[2*i+1] for i in range (m/2)]
    fe = fft (ae, m/2, w*w)
    fo = fft (ao, m/2, w*w)
    F = [0 for i in range(m)]
    x = 1
    for j in range (m/2):
        F[j] = fe[j] + x*fo[j]
        F[j+m/2] = fe[j] - x*fo[j]
        x *= w
    return F
    
def code():
    Q = raw_input()
    R = raw_input()
    Q = raw_input()
    S = raw_input()
    dlist1 = [int(x) for x in R.split(' ')]
    dlist2 = [int(x) for x in S.split(' ')]
    final = len (dlist1) + len (dlist2) - 1
    variable = pad(final)
    while len (dlist1) < variable:
        dlist1.append(0)
    while len (dlist2) < variable:
        dlist2.append(0)
    w = complex(math.cos(2*math.pi/variable), math.sin(2*math.pi/variable))
    newd1 = fft (dlist1, variable, w)
    newd2 = fft (dlist2, variable, w)
    outputlist = []
    for i in range(variable):
        outputlist.append(newd1[i]*newd2[i])
    P = fft(outputlist, variable, 1.0/w)
    print str(final-1)
    outputstr = ""
    for i in range(final):
        P[i] /= variable
        b = P[i].real + 0.5
        if b < 0:
            b -= 1
        outputstr += str(int(b)) + " "
    print outputstr[:-1]
        
    
N = int(raw_input())
while N > 0:
    code ()
    N -= 1
    
